crates/rome_js_analyze/src/analyzers/use_self_closing_elements.rs

```diff
@@ -0,0 +1,100 @@
+use rome_analyze::{ActionCategory, Rule, RuleCategory, RuleDiagnostic};
+use rome_console::markup;
+use rome_diagnostics::Applicability;
+use rome_js_factory::make;
+use rome_js_syntax::{JsAnyRoot, JsSyntaxToken, JsxAnyTag, JsxElement, JsxOpeningElementFields, T};
+use rome_rowan::{AstNode, AstNodeExt, AstNodeList, TriviaPiece};
+
+use crate::JsRuleAction;
+
+pub(crate) enum UseSelfClosingElements {}
+
+impl Rule for UseSelfClosingElements {
+    const NAME: &'static str = "useSelfClosingElements";
+    const CATEGORY: RuleCategory = RuleCategory::Lint;
+
+    type Query = JsxElement;
+    type State = ();
+
+    fn run(n: &Self::Query) -> Option<Self::State> {
+        if n.children().is_empty() {
+            Some(())
+        } else {
+            None
+        }
+    }
+
+    fn diagnostic(node: &Self::Query, _: &Self::State) -> Option<RuleDiagnostic> {
+        Some(RuleDiagnostic::warning(
+            node.range(),
+            markup! {
+                "JSX elements without children should be marked as self-closing. In JSX, it is valid for any element to be self-closing."
+            },
+        ))
+    }
+
+    fn action(root: JsAnyRoot, node: &Self::Query, _: &Self::State) -> Option<JsRuleAction> {
+        let open_element = node.opening_element().ok()?;
+        let JsxOpeningElementFields {
+            l_angle_token,
+            name,
+            type_arguments: _,
+            attributes,
+            r_angle_token,
+        } = open_element.as_fields();
+        let mut r_angle_token = r_angle_token.ok()?;
+        let mut leading_trivia = vec![];
+        let mut slash_token = String::new();
+
+        for trivia in r_angle_token.leading_trivia().pieces() {
+            leading_trivia.push(TriviaPiece::new(trivia.kind(), trivia.text_len()));
+            slash_token.push_str(trivia.text());
+        }
+        r_angle_token = r_angle_token.with_leading_trivia(std::iter::empty());
+        // check if previous `open_element` have a whitespace before `>`
+        // this step make sure we could convert <div></div> -> <div />
+        // <div test="some""></div> -> <div test="some" />
+        let need_extra_whitespace = if let Some(last_attribute) = attributes.last() {
+            let trailing = last_attribute.syntax().last_trailing_trivia();
+            if let Some(trailing) = trailing {
+                !trailing.text().ends_with(' ')
+            } else {
+                true
+            }
+        } else {
+            let name = name.clone().ok()?;
+            let trailing = name.syntax().last_trailing_trivia();
+            if let Some(trailing) = trailing {
+                !trailing.text().ends_with(' ')
+            } else {
+                true
+            }
+        };
```

Comment:
This could be simplified a bit by using `prev_token` to get to the token preceding `r_angle_token` directly (using the original version of the token, not the one with the trivias stripped) along with the `Option` combinator methods:
```suggestion
        let prev_token = r_angle_token.prev_token();
        let need_extra_whitespace = prev_token.as_ref()
            .map(|token| token.syntax())
            .and_then(|token| token.last_trailing_trivia())
            .map_or(true, |trailing| !trailing.text().ends_with(' '));
```

Reply:
I found that `r_angle_token.prev_token()` would always be `None`, so the `need_extra_whitespace` would always be true.

Reply:
That shouldn't be the case, are you calling `prev_token` on the token returned by `as_fields` ? The `with_leading_trivia` method returns a new detached token so it has no previous token, but the original token should definitely have a preceding token

Reply:
> 

You are right, Thanks!!

Reply:
done